# Customer Churn — Exploratory Data Analysis

## Project Objective

The goal of this project is to explore customer churn patterns using
Python, Pandas, and NumPy.

The analysis focuses on customer demographics, services, contract type,
payment methods, tenure, and charges to identify patterns associated
with customer churn.

The findings from this exploratory analysis can provide a foundation
for future customer churn prediction using machine learning.

In [1]:
import pandas as pd

In [2]:
import numpy as np

## 1. Dataset Loading & Initial Inspection

The dataset contains customer-level information including demographics,
services, contract details, payment methods, tenure, charges, and churn status.

In [3]:
data = pd.read_csv(r"Telco-Customer-Churn.csv")
data

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes


### Initial Dataset Structure

The dataset contains 7,043 customer records and 21 columns.

The `Churn` column will be treated as the primary outcome variable for
this exploratory analysis.

## 2. Data Quality Assessment

Before analyzing customer behavior, the dataset is checked for:

- Data types
- Missing values
- Duplicate customer records
- Potential data-format issues

In [5]:
data.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [8]:
data.customerID.duplicated().sum()

np.int64(0)

## 3. Descriptive Statistics

Descriptive statistics are used to understand the distribution of the
numerical variables, including customer tenure, monthly charges, and
total charges.

This provides an initial understanding of customer characteristics before
examining their relationship with churn.

In [23]:
data.describe()

,SeniorCitizen,tenure,MonthlyCharges,TotalCharges
count,7043.000000,7043.000000,7043.000000,7032.000000
mean,0.162147,32.371149,64.761692,2283.300441
std,0.368612,24.559481,30.090047,2266.771362
min,0.000000,0.000000,18.250000,18.800000
25%,0.000000,9.000000,35.500000,401.450000
50%,0.000000,29.000000,70.350000,1397.475000
75%,0.000000,55.000000,89.850000,3794.737500
max,1.000000,72.000000,118.750000,8684.800000


In [10]:
data2 = data.loc[data["MonthlyCharges"] > data["MonthlyCharges"].mean()]
data2

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.5,Yes
6,1452-KIOVK,Male,0,No,Yes,22,Yes,Yes,Fiber optic,No,...,No,No,Yes,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.4,No
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes
12,8091-TTVAX,Male,0,Yes,No,58,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,No,Credit card (automatic),100.35,5681.1,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7035,8456-QDAVC,Male,0,No,No,19,Yes,No,Fiber optic,No,...,No,No,Yes,No,Month-to-month,Yes,Bank transfer (automatic),78.70,1495.1,No
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes


In [11]:
data2.shape[0]

3923

## 3. Monthly Charges Analysis

Monthly charges are converted into a NumPy array to perform numerical
statistical analysis.

In [16]:
charges = data["MonthlyCharges"].to_numpy()
np.mean(charges)

np.float64(64.76169246059918)

In [13]:
np.median(charges)

np.float64(70.35)

In [14]:
np.std(charges)

np.float64(30.087910854936975)

In [15]:
filtered_data = data.loc[data['MonthlyCharges'] > 100]

In [16]:
filtered_data.shape[0]

902

In [17]:
charges_btw_50_100 = charges[(50 < charges) & (charges < 100)]
charges_btw_50_100

array([56.95, 53.85, 70.7 , ..., 60.65, 84.8 , 74.4 ], shape=(3836,))

In [18]:
charges_btw_50_100.size

3836

In [19]:
charges_above_100 = charges[charges > 100] 
charges_above_100.size

902

In [20]:
np.max(charges)

np.float64(118.75)

In [21]:
np.min(charges)

np.float64(18.25)

## 4. Understanding Customer Churn

The `Churn` variable indicates whether a customer left the company.

Before analyzing individual customer characteristics, the overall churn
distribution is examined to understand the proportion of customers who
churned versus those who remained.

In [24]:
data['Churn'].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [25]:
data["Contract"].value_counts()

Contract
Month-to-month    3875
Two year          1695
One year          1473
Name: count, dtype: int64

### Churn by Contract Type

Churn rates are compared across different contract types to determine
whether the length and structure of a customer's contract are associated
with different churn behavior.

In [63]:
churned_customers = data.loc[data['Churn'] == 'Yes'].groupby(['Contract'])['customerID'].count()
churned_customers

Contract
Month-to-month    1655
One year           166
Two year            48
Name: customerID, dtype: int64

In [65]:
total_customers = data.groupby('Contract')['customerID'].count()
total_customers

Contract
Month-to-month    3875
One year          1473
Two year          1695
Name: customerID, dtype: int64

In [66]:
churned_rate = churned_customers / total_customers * 100
churned_rate

Contract
Month-to-month    42.709677
One year          11.269518
Two year           2.831858
Name: customerID, dtype: float64

In [68]:
data['InternetService'].value_counts()

InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

In [70]:
internet_services = data.loc[data['Churn'] == 'Yes'].groupby(['InternetService'])['customerID'].count()
internet_services

InternetService
DSL             459
Fiber optic    1297
No              113
Name: customerID, dtype: int64

In [72]:
total_customer_by_internet = data.groupby(['InternetService'])['customerID'].count()
total_customer_by_internet

InternetService
DSL            2421
Fiber optic    3096
No             1526
Name: customerID, dtype: int64

In [73]:
churn_rate_by_internet = internet_services / total_customer_by_internet * 100
churn_rate_by_internet

InternetService
DSL            18.959108
Fiber optic    41.892765
No              7.404980
Name: customerID, dtype: float64

In [97]:
data.groupby('Churn')['MonthlyCharges'].mean()

Churn
No     61.265124
Yes    74.441332
Name: MonthlyCharges, dtype: float64

In [99]:
data.groupby('Churn')['tenure'].mean()

Churn
No     37.569965
Yes    17.979133
Name: tenure, dtype: float64

### Data Type Correction

`TotalCharges` is initially stored as a non-numeric data type.
Since it represents a monetary value, it needs to be converted to a
numeric format before statistical analysis.

In [103]:
data['TotalCharges'].dtype

dtype('O')

In [14]:
data['TotalCharges'] = pd.to_numeric(data['TotalCharges'], errors='coerce')

In [13]:
data['TotalCharges'].isnull().sum()

np.int64(0)

In [106]:
data.groupby('Churn')['TotalCharges'].mean()

Churn
No     2555.344141
Yes    1531.796094
Name: TotalCharges, dtype: float64

The `TotalCharges` column has been converted to a numeric format,
allowing it to be used for statistical analysis and comparison.

In [107]:
data.groupby('Contract')['MonthlyCharges'].mean()

Contract
Month-to-month    66.398490
One year          65.048608
Two year          60.770413
Name: MonthlyCharges, dtype: float64

In [108]:
data.groupby('Contract')['TotalCharges'].mean()

Contract
Month-to-month    1369.254581
One year          3034.683084
Two year          3728.933947
Name: TotalCharges, dtype: float64

In [110]:
data.groupby('Contract')['tenure'].mean()

Contract
Month-to-month    18.036645
One year          42.044807
Two year          56.735103
Name: tenure, dtype: float64

In [115]:
tenure_arr = data['tenure'].to_numpy()
tenure_arr[tenure_arr < 12].size

2069

### Tenure and Churn

Customer tenure is compared between churned and retained customers to
determine whether the length of the customer relationship is associated
with churn.

In [121]:
tenure_churn = data.loc[(data['tenure'] < 12) & (data['Churn'] == 'Yes')]['customerID'].count()
tenure_churn

np.int64(999)

In [126]:
total_churn = data.loc[data['tenure'] < 12]['customerID'].count()
total_churn

np.int64(2069)

In [128]:
churned_tenure = tenure_churn / total_churn * 100
churned_tenure

np.float64(48.28419526341228)

In [130]:
tenure_churn_12 = data.loc[(data['tenure'] >= 12) & (data['Churn'] == 'Yes')]['customerID'].count()
tenure_churn_12

np.int64(870)

In [131]:
total_churned = data.loc[data['tenure'] >= 12]['customerID'].count()
total_churned

np.int64(4974)

In [133]:
tenure_12 = tenure_churn_12 / total_churned * 100
tenure_12

np.float64(17.490952955367913)

Churned customers have substantially lower average tenure than customers
who remained, indicating that newer customers may be more vulnerable to churn.

## 5. Churn Rate Analysis by Customer Characteristics

Churn rates are compared across different customer characteristics
to identify groups with noticeably different churn behavior.

The analysis focuses on:

- Payment method
- Monthly Charges
- Internet service
- Senior citizen status
- Partner status
- Dependents

### Churn by Payment Method

Churn rates are compared across payment methods to identify whether
customer payment behavior is associated with differences in churn.

In [136]:
churned_by_payment = data.loc[data['Churn'] == 'Yes'].groupby('PaymentMethod')['customerID'].count()
churned_by_payment

PaymentMethod
Bank transfer (automatic)     258
Credit card (automatic)       232
Electronic check             1071
Mailed check                  308
Name: customerID, dtype: int64

In [137]:
total_customers_by_payment = data.groupby('PaymentMethod')['customerID'].count()
total_customers_by_payment

PaymentMethod
Bank transfer (automatic)    1544
Credit card (automatic)      1522
Electronic check             2365
Mailed check                 1612
Name: customerID, dtype: int64

In [138]:
churned_customers_by_payments =  churned_by_payment / total_customers_by_payment * 100
churned_customers_by_payments

PaymentMethod
Bank transfer (automatic)    16.709845
Credit card (automatic)      15.243101
Electronic check             45.285412
Mailed check                 19.106700
Name: customerID, dtype: float64

### Churn by Internet Service

Churn rates are compared across internet service types to identify
differences in churn behavior between customers using different services.

In [140]:
churned_by_internet = data.loc[data['Churn'] == 'Yes'].groupby('InternetService')['customerID'].count()
churned_by_internet

InternetService
DSL             459
Fiber optic    1297
No              113
Name: customerID, dtype: int64

In [141]:
total_by_internet = data.groupby('InternetService')['customerID'].count()
total_by_internet

InternetService
DSL            2421
Fiber optic    3096
No             1526
Name: customerID, dtype: int64

In [142]:
churned_customers_by_internet = churned_by_internet / total_by_internet * 100
churned_customers_by_internet

InternetService
DSL            18.959108
Fiber optic    41.892765
No              7.404980
Name: customerID, dtype: float64

### Monthly Charges and Churn

Average monthly charges are compared between customers who churned and
customers who remained to identify whether pricing levels are associated
with churn.

In [146]:
monthly_lessthan_50 = data.loc[(data['MonthlyCharges'] < 50) & (data['Churn'] == 'Yes')]['customerID'].count()
monthly_lessthan_50

np.int64(361)

In [151]:
total_monthly_50 = data.loc[data['MonthlyCharges'] < 50]['customerID'].count()
total_monthly_50

np.int64(2294)

In [154]:
monthly_churn_50 = monthly_lessthan_50 / total_monthly_50 * 100
monthly_churn_50

np.float64(15.736704446381866)

In [147]:
monthly_lessthan_100 = data.loc[(data['MonthlyCharges'] >= 50) & (data['MonthlyCharges'] < 100) & (data['Churn'] == 'Yes')]['customerID'].count()
monthly_lessthan_100

np.int64(1251)

In [152]:
total_monthly_100 = data.loc[(data['MonthlyCharges'] >= 50) & (data['MonthlyCharges'] < 100)]['customerID'].count()
total_monthly_100

np.int64(3841)

In [155]:
monthly_churn_100 = monthly_lessthan_100 / total_monthly_100 * 100
monthly_churn_100

np.float64(32.56964332205155)

In [159]:
monthly_greaterthan_100 = data.loc[(data['MonthlyCharges'] >=  100) & (data['Churn'] == 'Yes')]['customerID'].count()
monthly_greaterthan_100

np.int64(257)

In [160]:
total_monthly_greater_100 = data.loc[data['MonthlyCharges'] >= 100]['customerID'].count()
total_monthly_greater_100

np.int64(908)

In [161]:
monthly_churn_morethan_100 = monthly_greaterthan_100 / total_monthly_greater_100 * 100
monthly_churn_morethan_100

np.float64(28.30396475770925)

Churned customers have a higher average monthly charge than customers
who remained, suggesting an association between monthly charges and churn.

### Churn by Senior Citizen Status

Churn rates are compared between senior citizens and non-senior citizens
to determine whether age-related differences are associated with churn.

In [164]:
churn_by_senior = data.loc[data['Churn'] == 'Yes'].groupby('SeniorCitizen')['customerID'].count()
churn_by_senior

SeniorCitizen
0    1393
1     476
Name: customerID, dtype: int64

In [166]:
total_senior = data.groupby('SeniorCitizen')['customerID'].count()
total_senior

SeniorCitizen
0    5901
1    1142
Name: customerID, dtype: int64

In [167]:
churn_rate_by_senior = churn_by_senior / total_senior * 100
churn_rate_by_senior

SeniorCitizen
0    23.606168
1    41.681261
Name: customerID, dtype: float64

### Churn by Partner Status

Churn rates are compared between customers with and without a partner
to examine whether household relationship status is associated with churn.

In [7]:
churn_by_partner = data.loc[data['Churn'] == 'Yes'].groupby('Partner')['customerID'].count()
churn_by_partner

Partner
No     1200
Yes     669
Name: customerID, dtype: int64

In [6]:
partner_groupby = data.groupby('Partner')['customerID'].count()
partner_groupby

Partner
No     3641
Yes    3402
Name: customerID, dtype: int64

In [8]:
partner_churn_rate = churn_by_partner / partner_groupby * 100
partner_churn_rate

Partner
No     32.957979
Yes    19.664903
Name: customerID, dtype: float64

### Churn by Dependents

Churn rates are compared between customers with and without dependents
to examine whether dependent status is associated with churn.

In [9]:
churn_by_dependents = data.loc[data['Churn'] == 'Yes'].groupby('Dependents')['customerID'].count()
churn_by_dependents

Dependents
No     1543
Yes     326
Name: customerID, dtype: int64

In [11]:
dependents_groupby = data.groupby('Dependents')['customerID'].count()
dependents_groupby

Dependents
No     4933
Yes    2110
Name: customerID, dtype: int64

In [12]:
dependents_churn_rate = churn_by_dependents / dependents_groupby * 100
dependents_churn_rate

Dependents
No     31.279140
Yes    15.450237
Name: customerID, dtype: float64

## 7. Conclusion

In this project, I explored customer churn using Pandas and NumPy.

I looked at different factors such as tenure, monthly charges, contract
type, payment method, internet service, and customer details to understand
which groups had higher churn rates.

Some clear patterns were found. Customers with less than 12 months of
tenure had a much higher churn rate of 48.29% compared to 17.49% for
customers with longer tenure. Month-to-month customers also had a much
higher churn rate than customers with one-year or two-year contracts.

Electronic check and fiber optic customers also showed higher churn rates
compared to other groups. Senior citizens also had a higher churn rate
than non-senior customers.

Overall, this analysis helped me understand the main patterns in customer
churn and gave me a better understanding of how Pandas and NumPy can be
used to analyze real-world data.

This project can be taken further by using machine learning to predict
which customers are more likely to churn.